### ISSUE TITLE
networkd crashes when advertising blackhole route thru BGP

### ISSUE THREAD

**Issue by matijavizintin** [systemd/systemd#19648]
**systemd version the issue has been seen with**
 > systemd 245 (245.4-4ubuntu3.6)

**Used distribution**
 > Ubuntu 20.04.2 LTS

**Linux kernel version used** (`uname -a`)
 > `Linux hostname 5.4.0-73-generic #82-Ubuntu SMP Wed Apr 14 17:39:42 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > amd64

**Expected behaviour you didn't see**
 > networkd normally restarting 

**Unexpected behaviour you saw**
 > networkd crashes when a blackhole route is advertised thru BGP

**Steps to reproduce the problem**
 > install [FRR](https://deb.frrouting.org/) (stable)

 > /etc/frr/frr.conf
```
hostname localhost
ip forwarding
ipv6 forwarding
line vty
service integrated-vtysh-config
router bgp 64000
 bgp router-id <IP>
 network <IP>/32
 aggregate-address <NETWORK_ADDRESS>/16
 !
 neighbor mypeer peer-group
 neighbor mypeer remote-as external
 !
 neighbor <INTERFACE> interface v6only peer-group mypeer
```

> /etc/frr/daemons
```
bgpd=yes
```

> systemctl restart frr
> systemctl restart systemd-networkd

**Additional program output to the terminal or log subsystem illustrating the issue**
> ip r | grep blackhole
```
blackhole 10.0.0.0/16 proto bgp metric 20
```
> journalctl -u systemd-networkd
```
May 18 12:29:28 routing-focal-l3 systemd[1]: Starting Network Service...
May 18 12:29:28 routing-focal-l3 systemd-networkd[16485]: enp0s3: Gained IPv6LL
May 18 12:29:28 routing-focal-l3 systemd-networkd[16485]: Assertion 'ifindex' failed at src/network/networkd-link.c:757, function link_get(). Aborting.
May 18 12:29:28 routing-focal-l3 systemd[1]: systemd-networkd.service: Main process exited, code=dumped, status=6/ABRT
May 18 12:29:28 routing-focal-l3 systemd[1]: systemd-networkd.service: Failed with result 'core-dump'.
May 18 12:29:28 routing-focal-l3 systemd[1]: Failed to start Network Service.
May 18 12:29:28 routing-focal-l3 systemd[1]: systemd-networkd.service: Scheduled restart job, restart counter is at 1.
May 18 12:29:28 routing-focal-l3 systemd[1]: Stopped Network Service.
```

**Comment by poettering** [comment]
*[Canned reply follows]*

This is the upstream bug and feature request tracker of systemd. Please use this only for issues in the two most current upstream systemd versions. See this link for the list of current releases.

https://github.com/systemd/systemd/releases

For support for older versions please contact your distribution instead.

In order to keep frustration at a minimum the submission form tried to make this all very clear.

**Comment by yuwata** [comment]
I am not familiar with that tool. Could you try to reproduce the issue without frr and with one of the recent systemd releases? Then, still the recent systemd has the same issue, please provide relevant .network (and .netdev if you use) files and debugging logs of networkd. You can generate debugging logs by creating the following drop-in file.
```
# /etc/systemd/system/systemd-networkd.service.d/override.conf
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
```

**Comment by matijavizintin** [comment]
Hey @yuwata 

[FRR](https://github.com/FRRouting/frr) is a routing daemon, a fork from [quagga](https://github.com/Quagga/quagga). If FRR is stopped then networkd works fine although i add the static blackhole route (ip route add blackhole 10.0.0.0/16). It seems that the bgp route is installed somehow differently. 

I'll try with the latest systemd, in the mean time here is a debug log from networkd.

Here is a debug log from a test VM:
```
May 18 17:06:33 routing-focal-l3 systemd[1]: Starting Network Service...
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: Bus bus-api-network: changing state UNSET → OPENING
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: Bus bus-api-network: changing state OPENING → AUTHENTICATING
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: timestamp of '/etc/systemd/network' changed
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: timestamp of '/run/systemd/network' changed
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: Virtualization innotek GmbH found in DMI (/sys/class/dmi/id/sys_vendor)
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: Found VM virtualization oracle
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: /usr/lib/systemd/network/80-container-host0.network: Conditions in the file do not match the system environment, skipping.
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy4: New device has no master, continuing without
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy4: Flags change: +BROADCAST +NOARP
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy4: Link 6 added
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy4: udev initialized link
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy4: Saved original MTU: 1500
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy3: New device has no master, continuing without
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy3: Flags change: +BROADCAST +NOARP
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy3: Link 5 added
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy3: udev initialized link
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy3: Saved original MTU: 1500
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy2: New device has no master, continuing without
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy2: Flags change: +BROADCAST +NOARP
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy2: Link 4 added
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy2: udev initialized link
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy2: Saved original MTU: 1500
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy1: New device has no master, continuing without
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy1: Flags change: +BROADCAST +NOARP
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy1: Link 3 added
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy1: udev initialized link
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: dummy1: Saved original MTU: 1500
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: New device has no master, continuing without
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Flags change: +UP +LOWER_UP +RUNNING +MULTICAST +BROADCAST
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Link 2 added
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: udev initialized link
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Saved original MTU: 1500
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: New device has no master, continuing without
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Flags change: +LOOPBACK +UP +LOWER_UP +RUNNING
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Link 1 added
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: udev initialized link
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Saved original MTU: 65536
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Remembering foreign address: fe80::a4:b5ff:feb1:da19/64 (valid forever)
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Gained IPv6LL
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Remembering foreign address: ::1/128 (valid forever)
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Remembering updated address: 10.0.2.15/24 (valid for 23h 57min 26s)
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Remembering foreign address: 10.48.0.10/32 (valid forever)
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Remembering foreign address: 127.0.0.1/8 (valid forever)
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: rtnl: received non-static neighbor, ignoring.
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: rtnl: received non-static neighbor, ignoring.
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: rtnl: received non-static neighbor, ignoring.
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: rtnl: received non-static neighbor, ignoring.
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: rtnl: received non-static neighbor, ignoring.
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: rtnl: received non-static neighbor, ignoring.
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: rtnl: received non-static neighbor, ignoring.
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: rtnl: received non-static neighbor, ignoring.
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: rtnl: received non-static neighbor, ignoring.
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Remembering route: dst: ff00::/8, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local, proto: boot, type: unicast
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Remembering route: dst: fe80::a4:b5ff:feb1:da19/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local, proto: kernel, type: local
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Remembering route: dst: fe80::/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local, proto: kernel, type: anycast
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Remembering route: dst: ::1/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: local, proto: kernel, type: local
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Remembering route: dst: fe80::/64, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: kernel, type: unicast
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Remembering route: dst: ::1/128, src: n/a, gw: n/a, prefsrc: n/a, scope: global, table: main, proto: kernel, type: unicast
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Remembering route: dst: 127.255.255.255/32, src: n/a, gw: n/a, prefsrc: 127.0.0.1, scope: link, table: local, proto: kernel, type: broadcast
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Remembering route: dst: 127.0.0.1/32, src: n/a, gw: n/a, prefsrc: 127.0.0.1, scope: host, table: local, proto: kernel, type: local
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Remembering route: dst: 127.0.0.0/8, src: n/a, gw: n/a, prefsrc: 127.0.0.1, scope: host, table: local, proto: kernel, type: local
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Remembering route: dst: 127.0.0.0/32, src: n/a, gw: n/a, prefsrc: 127.0.0.1, scope: link, table: local, proto: kernel, type: broadcast
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: lo: Remembering route: dst: 10.48.0.10/32, src: n/a, gw: n/a, prefsrc: 10.48.0.10, scope: host, table: local, proto: kernel, type: local
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Remembering route: dst: 10.0.2.255/32, src: n/a, gw: n/a, prefsrc: 10.0.2.15, scope: link, table: local, proto: kernel, type: broadcast
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Remembering route: dst: 10.0.2.15/32, src: n/a, gw: n/a, prefsrc: 10.0.2.15, scope: host, table: local, proto: kernel, type: local
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Remembering route: dst: 10.0.2.0/32, src: n/a, gw: n/a, prefsrc: 10.0.2.15, scope: link, table: local, proto: kernel, type: broadcast
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Remembering route: dst: 10.0.2.2/32, src: n/a, gw: n/a, prefsrc: 10.0.2.15, scope: link, table: main, proto: dhcp, type: unicast
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Remembering route: dst: 10.0.2.0/24, src: n/a, gw: n/a, prefsrc: 10.0.2.15, scope: link, table: main, proto: kernel, type: unicast
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: rtnl: received route message without ifindex, ignoring
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: enp0s3: Remembering route: dst: n/a, src: n/a, gw: 10.0.2.2, prefsrc: 10.0.2.15, scope: global, table: main, proto: dhcp, type: unicast
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: rtnl: received rule message with invalid family 129, ignoring.
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: rtnl: received rule message with invalid family 128, ignoring.
May 18 17:06:34 routing-focal-l3 systemd-networkd[19158]: Assertion 'ifindex' failed at src/network/networkd-link.c:757, function link_get(). Aborting.
May 18 17:06:34 routing-focal-l3 systemd[1]: systemd-networkd.service: Main process exited, code=dumped, status=6/ABRT
May 18 17:06:34 routing-focal-l3 systemd[1]: systemd-networkd.service: Failed with result 'core-dump'.
May 18 17:06:34 routing-focal-l3 systemd[1]: Failed to start Network Service.
May 18 17:06:34 routing-focal-l3 systemd[1]: systemd-networkd.service: Scheduled restart job, restart counter is at 2.
May 18 17:06:34 routing-focal-l3 systemd[1]: Stopped Network Service.
```

**Comment by yuwata** [comment]
Please also provide coredump. `link_get()` is called at so many places. I'd like to know which call triggers the assertion.

**Comment by matijavizintin** [comment]
Here is the [coredump](https://drive.google.com/file/d/1tCKiOSnGkLbx3w9du5_NNepkOQZC_hx7/view?usp=sharing)

And here is the .network file (vagrant autogenerated)
```

root@routing-focal-l3:/run/systemd/network# cat 10-netplan-enp0s3.network
[Match]
MACAddress=02:a4:b5:b1:da:19
Name=enp0s3

[Network]
DHCP=ipv4
LinkLocalAddressing=ipv6

[DHCP]
RouteMetric=100
UseMTU=true
```

**Comment by ddstreet** [comment]
The bt is:

> (gdb) bt
> #0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
> #1  0x00007ffa7353e859 in __GI_abort () at abort.c:79
> #2  0x00005622e3199d3a in log_assert_failed_realm (realm=<optimized out>, text=0x5622e32ee275 "ifindex", file=0x5622e330e993 "src/network/networkd-link.c", line=757, func=0x5622e3314988 <__PRETTY_FUNCTION__.22001> "link_get") at ../src/basic/log.c:809
> #3  0x00005622e32958a1 in link_get (m=<optimized out>, ifindex=<optimized out>, ret=<optimized out>) at ../src/network/networkd-link.c:757
> #4  0x00005622e32d56f8 in manager_rtnl_process_nexthop (rtnl=<optimized out>, message=0x5622e343c610, userdata=0x5622e341ec30) at ../src/network/networkd-manager.c:1265
> #5  0x00005622e319b8c4 in manager_rtnl_enumerate_nexthop (m=0x5622e341ec30) at ../src/network/networkd-manager.c:2118
> #6  run (argv=<optimized out>, argc=<optimized out>) at ../src/network/networkd.c:111
> #7  main (argc=<optimized out>, argv=<optimized out>) at ../src/network/networkd.c:130
> 

It looks like this is already fixed by commit 25b831bac8a5e545e1eda5199392c11c7aed4e42 so I'll backport that into ubuntu.
